src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -318,13 +347,99 @@ def _prepare_inits(self) -> Set[str]:
   def set_emit_warnings(self, emit_warnings):
     self._builder.info.emit_warnings = emit_warnings
 
+  def _shuffle_original_build_info_into_ptex(self):
+    """Create a "dehydrated" ptex file without any of its requirements, and specify that in two *-INFO files.
+
+    See this test case as background:
+    https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R23
+    """
+    ipex_info = self._builder.info.copy()
+    orig_chroot = self._builder.chroot()
+    # Replace the original builder with a new one, and just pull files from the old chroot.
+    self._builder = PEXBuilder(interpreter=self._builder.interpreter)
+
+    # If the dependencies are very large, this flag is indicated to be useful in RAM-constrained
+    # environments.
+    self._builder.info.always_write_cache = True
+
+    # PTEX-INFO: Contains resolver settings and source files to copy over into the dehydrated .ipex
+    #            file.
+    all_code = []
+    for label in ['source', 'resource']:
+      for src in orig_chroot.get(label):
+        if src in ['__main__.py', PexInfo.PATH]:
```

Comment:
I'm fine with this, and this is code-golfy, but you are intercepting all calls to your underlying `_builder` and already know all the source and resource files added. You need not use these magic (private!) PEXBuilder chroot labels and also filter out `__main__.py` and `PEX-INFO`, you can just directly save the sources and resources paths like you do `_distributions` and then access those lists here directly with no filtering.